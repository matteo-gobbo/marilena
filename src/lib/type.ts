import { FastifyViewOptions } from "@fastify/view";

export interface Config {
  templateSuffix: string; // .html .eta .handlebars ...etc
  inputFolder: string; // default input
  outputFolder: string; // default output
  engine?: keyof FastifyViewOptions["engine"];
  variablesType: "json"; // json or yml in the future
  locales: string[]; // default ["en"]
}
