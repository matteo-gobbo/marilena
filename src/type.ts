export interface Options {
    inoputFolder: string
    outputFolder: string
}