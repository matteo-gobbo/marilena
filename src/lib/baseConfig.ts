import { Config } from "../types";

const config: Config = {
  templateSuffix: ".html",
  inputFolder: "input",
  outputFolder: "output",
  locales: ["en"],
};

export default config;
