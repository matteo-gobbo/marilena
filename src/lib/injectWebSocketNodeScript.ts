import { parse } from "node-html-parser";
import { EVENT_NAME_NEED_REFRESH_WEBSOCKET } from "../const";

// this method append before body script mandatory to catch refresh by server via websocket
export function inject(html: string): string {
  const root = parse(html);

  const scriptString = `
    <script>
    const socket = new WebSocket("ws://localhost:8080/socket");
  
    // Listen for messages
    socket.addEventListener("message", (event) => {
      const message = event.data;
      if(message === "${EVENT_NAME_NEED_REFRESH_WEBSOCKET}"){
        window.location.reload();
      }
    });
    </script>
    `;

  root.insertAdjacentHTML("beforeend", scriptString);
  return root.innerHTML;
}
