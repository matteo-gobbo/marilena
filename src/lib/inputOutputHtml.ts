import mjml2html from "mjml";
import { Config } from "../types";
import eta from "eta";
import logger from "node-color-log";
import { CONFIG_FILE_NAME } from "../const";

const isTextExecution = process.env.NODE_ENV === "test";
// this method should take soem html, add variables, convert with mjml and return html

interface Options {
  templateOptions?: Config["templateOptions"];
  inputHtml: string;
  variables: object;
}
export function inputOutputHtml({
  inputHtml,
  variables,
  templateOptions,
}: Options): string {
  function rendereWithVars() {
    if (!templateOptions) {
      return inputHtml;
    }

    // eta.templates.define("aaa", eta.compile("PARTIAL"));

    switch (templateOptions.engine) {
      case "eta": {
        const { prepareEngine } = templateOptions;
        if (!!prepareEngine) {
          prepareEngine(eta);
        } else {
          logger.error(
            `templateOptions options is defined, but prepareEngine is null. Please check method under ${CONFIG_FILE_NAME}`
          );
        }

        return eta.render(inputHtml, variables);
      }

      default: {
        logger.info(
          `engine ${templateOptions.engine} not supported. Please contribute to the repo :)`
        );
        logger.info("template will be parsed without fill variables");
        return inputHtml;
      }
    }
  }

  const mjmlTenplateWithVars = rendereWithVars();

  if (isTextExecution) {
    return mjmlTenplateWithVars;
  }

  const html = mjml2html(mjmlTenplateWithVars).html;
  return html;
}
