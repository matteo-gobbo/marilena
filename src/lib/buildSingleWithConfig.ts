import { removeExtension } from "../utils";
import fs from "fs";
import path from "path";
import { Config } from "./type";
import { inputOutputHtml } from "./inputOutputHtml";
import loadVariables from "./loadVariables";
import logger from "node-color-log";

export function buildSingle(config: Config, emailName: string) {
  const { inputFolder, outputFolder, locales } = config;

  const inputFolderPath = path.join(inputFolder);
  const outputFolderPath = path.join(outputFolder);

  const inputEmailFilePath = path.join(
    inputFolderPath,
    emailName,
    "index.html"
  );
  const mjmlTemplate = fs.readFileSync(inputEmailFilePath, "utf-8");

  // create folder for each email and locale if not exist
  locales.forEach((locale) => {
    const emailFolderPath = path.join(outputFolderPath, emailName, locale);
    if (!fs.existsSync(emailFolderPath)) {
      fs.mkdirSync(emailFolderPath, { recursive: true });
    }

    const variables = loadVariables({ config, locale, emailName });
    const html = inputOutputHtml({
      inputHtml: mjmlTemplate,
      variables,
      templateOptions: config.templateOptions,
    });

    const folderEmailPathLang = path.join(outputFolderPath, emailName, locale);
    if (!fs.existsSync(folderEmailPathLang)) {
      fs.mkdirSync(folderEmailPathLang, { recursive: true });
    }

    try {
      fs.writeFileSync(path.join(folderEmailPathLang, "index.html"), html, {
        encoding: "utf-8",
      });
      logger.info("writed", emailName, locale);
    } catch (e) {
      logger.error("failed writing", emailName, e);
    }
  });
}
