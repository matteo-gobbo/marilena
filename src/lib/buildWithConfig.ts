import { removeExtension } from "../utils";
import fs from "fs";
import path from "path";
import { Config } from "../types";
import { inputOutputHtml } from "./inputOutputHtml";
import { buildSingle } from "./buildSingleWithConfig";
import loadVariables from "./loadVariables";
import logger from "node-color-log";

export function build(config: Config) {
  const { inputFolder, outputFolder, templateSuffix, locales } = config;

  const inputFolderPath = path.join(inputFolder);
  const outputFolderPath = path.join(outputFolder);

  // create folder if not exist
  if (!fs.existsSync(outputFolderPath)) {
    fs.mkdirSync(outputFolderPath, { recursive: true });
  }

  let list: string[];
  try {
    const files = fs.readdirSync(inputFolderPath, { withFileTypes: true });
    list = files.filter((f) => f.isDirectory()).map((f) => f.name);
  } catch (e) {
    console.error(e);
    list = [];
  }

  list.forEach((emailName) => {
    buildSingle(config, emailName);
  });
}
