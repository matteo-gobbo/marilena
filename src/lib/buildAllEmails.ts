import { loadConfig } from "../utils";
import { build } from "./buildWithConfig";

async function buildAllEmail() {
  const config = await loadConfig();
  build(config);
}

buildAllEmail();
