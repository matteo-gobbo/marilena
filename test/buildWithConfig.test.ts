import { describe, expect, test } from "@jest/globals";
import { Config } from "../src/lib/type";
import { build } from "../src/lib/buildWithConfig";
import fs from "fs";
import path from "path";

const testEmailName = "hello"; // we have only hello.html inside inputs

describe("writer correctly email files", () => {
  test("create email with: [engine: NO, suffix: html, locale: en]", () => {
    const config: Config = {
      templateSuffix: ".html",
      inputFolder: "test/basic_1/input",
      outputFolder: "test/basic_1/output",
      locales: ["en"],
    };

    build(config);

    const outpoutEn = fs.readFileSync(
      path.join(config.outputFolder, testEmailName, "en", "index.html"),
      { encoding: "utf-8" }
    );

    fs.rmSync(config.outputFolder, { recursive: true });

    expect(outpoutEn).toMatchSnapshot();
  });

  test("create email with: [engine: eta, suffix: html, locale: en]", () => {
    const config: Config = {
      templateSuffix: ".html",
      inputFolder: "test/eta_1/input",
      outputFolder: "test/eta_1/output",
      locales: ["en"],
      templateOptions: {
        variablesType: "json",
        engine: "eta",
      },
    };

    build(config);

    const outpoutEn = fs.readFileSync(
      path.join(config.outputFolder, testEmailName, "en", "index.html"),
      { encoding: "utf-8" }
    );

    fs.rmSync(config.outputFolder, { recursive: true });

    expect(outpoutEn).toMatchSnapshot();
  });
});
