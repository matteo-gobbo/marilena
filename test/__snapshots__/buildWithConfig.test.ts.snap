// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writer correctly email files create email with: [engine: NO, suffix: html, locale: en] 1`] = `
"<mjml>
  <mj-body>
    <mj-section>
      <mj-column>
        <mj-text>test</mj-text>
      </mj-column>
    </mj-section>
  </mj-body>
</mjml>
"
`;

exports[`writer correctly email files create email with: [engine: eta, suffix: html, locale: en] 1`] = `
"<mjml>
  <mj-body>
    <mj-section>
      <mj-column>
        <mj-text>hello en</mj-text>
        <mj-text>this is partial 1</mj-text>
      </mj-column>
    </mj-section>
  </mj-body>
</mjml>
"
`;
